File: Stack/quicksort_iterative.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

// Partition procedure (Procedure 6.5)
int QUICK(vector<int> &A, int BEG, int END) {
    int LEFT = BEG, RIGHT = END, LOC = BEG;
    while (true) {
        // Scan from right to left
        while (A[LOC] <= A[RIGHT] && LOC != RIGHT)
            RIGHT--;
        if (LOC == RIGHT) break;
        if (A[LOC] > A[RIGHT]) {
            swap(A[LOC], A[RIGHT]);
            LOC = RIGHT;
        }

        // Scan from left to right
        while (A[LEFT] <= A[LOC] && LEFT != LOC)
            LEFT++;
        if (LOC == LEFT) break;
        if (A[LEFT] > A[LOC]) {
            swap(A[LEFT], A[LOC]);
            LOC = LEFT;
        }
    }
    return LOC;
}

// Non-recursive QuickSort (Algorithm 6.6)
void quickSort(vector<int> &A, int N) {
    stack<int> LOWER, UPPER;
    LOWER.push(0);
    UPPER.push(N - 1);

    while (!LOWER.empty()) {
        int BEG = LOWER.top(); LOWER.pop();
        int END = UPPER.top(); UPPER.pop();

        if (BEG < END) {
            int LOC = QUICK(A, BEG, END);

            // Push left sublist
            if (BEG < LOC - 1) {
                LOWER.push(BEG);
                UPPER.push(LOC - 1);
            }
            // Push right sublist
            if (LOC + 1 < END) {
                LOWER.push(LOC + 1);
                UPPER.push(END);
            }
        }
    }
}

int main() {
    vector<int> arr = {34, 7, 23, 32, 5, 62};
    int N = arr.size();

    quickSort(arr, N);

    cout << "Sorted array: ";
    for (int x : arr) cout << x << " ";
    cout << endl;

    return 0;
}

File: Stack/quicksort_iterative.py
def QUICK(A, BEG, END):
    LEFT, RIGHT, LOC = BEG, END, BEG
    while True:
        # Scan from right to left
        while A[LOC] <= A[RIGHT] and LOC != RIGHT:
            RIGHT -= 1
        if LOC == RIGHT:
            break
        if A[LOC] > A[RIGHT]:
            A[LOC], A[RIGHT] = A[RIGHT], A[LOC]
            LOC = RIGHT

        # Scan from left to right
        while A[LEFT] <= A[LOC] and LEFT != LOC:
            LEFT += 1
        if LOC == LEFT:
            break
        if A[LEFT] > A[LOC]:
            A[LEFT], A[LOC] = A[LOC], A[LEFT]
            LOC = LEFT
    return LOC


def quickSort(A):
    LOWER, UPPER = [], []
    LOWER.append(0)
    UPPER.append(len(A) - 1)

    while LOWER:
        BEG, END = LOWER.pop(), UPPER.pop()

        if BEG < END:
            LOC = QUICK(A, BEG, END)

            # Push left sublist
            if BEG < LOC - 1:
                LOWER.append(BEG)
                UPPER.append(LOC - 1)

            # Push right sublist
            if LOC + 1 < END:
                LOWER.append(LOC + 1)
                UPPER.append(END)


# Example usage
arr = [34, 7, 23, 32, 5, 62]
quickSort(arr)
print("Sorted array:", arr)
